File: PascalsTriangle/Triangle.cs
using System.Collections.Generic;
using System.Linq;
using static LanguageExt.Prelude;

namespace PascalsTriangle
{
    public static class Triangle
    {
        public static IEnumerable<uint> DrawTriangle(uint rows) =>
            Enumerable
                .Range(0, (int) rows)
                .Aggregate(EmptyList, BuildTriangle);

        private static IEnumerable<uint> BuildTriangle(IEnumerable<uint> triangle, int currentRowNumber)
        {
            return (triangle is null, currentRowNumber) switch
            {
                (true, _) => triangle,
                (_, 0) => FirstRow,
                (_, 1) => SecondRow,
                (_, _) => BuildNextRow(triangle, currentRowNumber)
            };

            static IEnumerable<uint> BuildNextRow(IEnumerable<uint> triangle, int previousRowNumber) =>
                triangle
                    .TakeLast(previousRowNumber)
                    .Pairwise()
                    .Aggregate(triangle.Append(One), (accumulator, pair) => accumulator.Append(pair.Previous + pair.Current))
                    .Append(One);
        }

        private const uint One = 1;
        private static IEnumerable<uint> EmptyList => List<uint>().AsEnumerable();
        private static IEnumerable<uint> FirstRow => List(One);
        private static IEnumerable<uint> SecondRow => List<uint>(1, 1, 1);
    }
}

File: PascalsTriangle/Triangle.fs
namespace Pascals

module Triangle =
    
    let private buildNewRow triangle previousRowNumber =
        let row = triangle
                    |> List.rev
                    |> List.take(previousRowNumber)
                    |> List.pairwise
                    |> List.fold (fun row (prev:int, current:int) -> row @ [prev + current]) [1]
                
        triangle @ row @ [1]

    let private buildTriangle rows =
        [2 .. rows-1]
        |> List.fold buildNewRow [1;1;1]

    let build rows =
        match rows with
        | x when x <= 0 -> []
        | 1 -> [1]
        | 2 -> [1;1;1]
        | _ -> buildTriangle rows
